src/server/auth/server/saml.go

```diff
@@ -24,43 +26,72 @@ import (
 type canonicalConfig struct {
 	Version int64
 
-	ACSURL      *url.URL
-	MetadataURL *url.URL
-	DashURL     *url.URL
+	// Currently, both IDP and SAMLSvc must be set. They are separate structs to
+	// separate ambiguous names (e.g. IDP.MetadataURL is the ID provider's
+	// metadata URL, while SAMLSvc.MetadataURL is Pachyderm's metadata URL,
+	// configured by the cluster admin.
+	//
+	// Later, there may be multiple IDPs, or IDP may be set to a non-SAML source
+	// and SAMLSvc may not be set, but those aren't supported yet
+	IDP struct {
+		Name           string
+		Description    string
+		MetadataURL    *url.URL
+		Metadata       *saml.EntityDescriptor
+		GroupAttribute string
+	}
 
-	IDPName        string
-	IDPDescription string
-	IDPMetadataURL *url.URL
-	IDPMetadata    *saml.EntityDescriptor
+	SAMLSvc struct {
+		ACSURL          *url.URL
+		MetadataURL     *url.URL
+		DashURL         *url.URL
+		SessionDuration time.Duration
+	}
 }
 
-func (c *canonicalConfig) ToProto() *auth.AuthConfig {
-	if c == nil {
-		return &auth.AuthConfig{}
+func (c *canonicalConfig) ToProto() (*auth.AuthConfig, error) {
```

Comment:
Is it infeasible to use a protocol buffer as the canonical config? Having to convert between a go struct and a protobuf go struct seems to lead to bugs where you add a field but forget to add the field to the `ToProto` method. Sometimes it's unavoidable, if you need non-data fields in the struct but just glancing at it seems like `canonicalConfig` might not have any of those.

Reply:
My thoughts on this depend on what you have in mind.

Basically, I like the idea of having separate data structures for configs and canonical configs, because I want to avoid migration pain (the `AuthConfig` proto is part of the `SetConfig` API, so my plan is not to remove stuff from that, and it's designed to be more flexible than what's actually supported, with multiple IDPs possible, or an IDP with no SAML settings).

That said, if keeping `canonicalConfig` is okay, and the question is whether it's a proto or struct—I'm fine with either, but it seems like either way I'll have to deal with conversion from `canonicalConfig` (which is internal and will change as I add stuff) to `AuthConfig` (which is public and won't change much). Does making it a proto make the conversion less error prone? Even if I can use reflection to copy fields by name, that kind of makes me nervous since the point of having two data structures is that one must be valid, so I might be setting myself up to forget validation logic that I actually need.

One idea that I kind of like, but would probably do later is to write a simple fuzzer that generates random configs and then sets & gets them to make sure fields aren't dropped